### PR TITLE
feat(eks): kubernetes resource pruning

### LINKED ISSUE #10495: [aws-eks] Support object pruning
### Use Case

Currently, situations may arise that leave orphan resources. For example, consider the following manifest:

```typescript
new eks.KubernetesManifest(this, 'Configs', {
  cluster,
  manifest: [
    {
      kind: 'ConfigMap',
      apiVersion: 'v1',
      metadata: {
        name: 'config1'
      },
      data: {
        foo: 'bar'
      }
    }, 
    {
      kind: 'ConfigMap',
      apiVersion: 'v1',
      metadata: {
        name: 'config2'
      },
      data: {
        hello: 'world'
      }
    }, 
  ]
})
```

When it is first deployed, we issue a `kubectl apply` command that creates both config maps.

If we now remove one config map, i.e the manifest will look like so:

```typescript
new eks.KubernetesManifest(this, 'Configs', {
  cluster,
  manifest: [
    {
      kind: 'ConfigMap',
      apiVersion: 'v1',
      metadata: {
        name: 'config1'
      },
      data: {
        foo: 'bar'
      }
    } 
  ]
})
```

Since the CFN resource is the same, just its properties changed, an `UPDATE` call be executed by CFN, which will simply translate into a `kubectl apply` on the new manifest, which will not delete the config map that was removed from the manifest.

### Proposed Solution

- Automatically add pruning labels to all resources and use `kubectl apply --prune` (see [`pruning objects`]( https://kubernetes.io/docs/tasks/manage-kubernetes-objects/declarative-config/#alternative-kubectl-apply-f-directory-prune-l-your-label))

- Before running `kubectl apply` on update, run `kubectl delete` on the previous manifest (available as as CFN property at invocation time)

### Other

* [X] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by eladb**
In order to support deletion of kubernetes manifest resources, the EKS module now automatically allocates and injects a "prune label" to all resources. This label is then passed down to `kubectl apply` with the `--prune` option so that any resources in the cluster that do not appear in the manifest will get deleted.

The `prune` option can be set to `false` (either at the `Cluster` level or at the KubernetesResource level) to disable this.

In order to avoid needing to update all tests, many of the existing tests set `prune: false` so that their outputs are not impacted.

Resolves #10495

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by eladb** [comment]
From the kubectl logs:

```
Running command: ['kubectl', 'apply', '--kubeconfig', '/tmp/kubeconfig', '-f', '/tmp/manifest.yaml', '--prune', '-l', 'aws.cdk.eks/prune-c8f0f7140f7358e29b7f58e81b507dcf744a3908f4']
b'deployment.apps/hello-kubernetes unchanged\nservice/hello-kubernetes pruned\n'
```

File: packages/@aws-cdk/aws-eks/lib/kubectl-handler/apply/__init__.py [hunk 1]
```diff
@@ -40,20 +41,23 @@ def apply_handler(event, context):
     logger.info("manifest written to: %s" % manifest_file)
 
     if request_type == 'Create' or request_type == 'Update':
-        kubectl('apply', manifest_file)
+        opts = []
+        if prune_label is not None:
+            opts = ['--prune', '-l', prune_label]
+        kubectl('apply', manifest_file, *opts)
```
Comment (ayush987goyal) [inline packages/@aws-cdk/aws-eks/lib/kubectl-handler/apply/__init__.py:47]: Thanks for this change! We can now address this issue in similar fashion: https://github.com/aws/aws-cdk/issues/11763

Comment (eladb) [reply packages/@aws-cdk/aws-eks/lib/kubectl-handler/apply/__init__.py:47]: > Thanks for this change! We can now address this issue in similar fashion: https://github.com/aws/aws-cdk/issues/11763

Yes. Sounds pretty straightforward to support. Contributions are welcome!

Comment (ayush987goyal) [reply packages/@aws-cdk/aws-eks/lib/kubectl-handler/apply/__init__.py:47]: Yeah. Planning to contribute post this change is merged!

**Review by iliapolo — APPROVED** [review]

**Pull request merged**